src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs

```diff
@@ -6723,5 +6723,72 @@ static void Main(string[] args)
+
+            Action<ModuleSymbol> validator = module =>
+            {
+                var typeParameter = module.GlobalNamespace.GetTypeMember("Test").TypeParameters.Single();
+                Assert.False(typeParameter.HasValueTypeConstraint);
+                Assert.False(typeParameter.HasReferenceTypeConstraint);
+                Assert.Equal(SpecialType.System_Enum, typeParameter.ConstraintTypes().Single().SpecialType);
+            };
+
+            CompileAndVerify(code, sourceSymbolValidator: validator, symbolValidator: validator);
+        }
+
+        [Fact]
+        public void EnumConstraintIsReflectedinSymbols_ValueType()
+        {
+            var code = "public class Test<T> where T : struct, System.Enum { }";
+
+            Action<ModuleSymbol> validator = module =>
+            {
+                var typeParameter = module.GlobalNamespace.GetTypeMember("Test").TypeParameters.Single();
+                Assert.True(typeParameter.HasValueTypeConstraint);
+                Assert.False(typeParameter.HasReferenceTypeConstraint);
+                Assert.False(typeParameter.HasConstructorConstraint);
+                Assert.Equal(SpecialType.System_Enum, typeParameter.ConstraintTypes().Single().SpecialType);
+            };
+
+            CompileAndVerify(code, sourceSymbolValidator: validator, symbolValidator: validator);
+        }
+
+        [Fact]
+        public void EnumConstraintIsReflectedinSymbols_ReferenceType()
+        {
+            var code = "public class Test<T> where T : class, System.Enum { }";
+
+            Action<ModuleSymbol> validator = module =>
+            {
+                var typeParameter = module.GlobalNamespace.GetTypeMember("Test").TypeParameters.Single();
+                Assert.False(typeParameter.HasValueTypeConstraint);
+                Assert.True(typeParameter.HasReferenceTypeConstraint);
+                Assert.False(typeParameter.HasConstructorConstraint);
+                Assert.Equal(SpecialType.System_Enum, typeParameter.ConstraintTypes().Single().SpecialType);
+            };
+
+            CompileAndVerify(code, sourceSymbolValidator: validator, symbolValidator: validator);
+        }
+
+        [Fact]
+        public void EnumConstraintIsReflectedinSymbols_Constructor()
+        {
+            var code = "public class Test<T> where T : System.Enum, new() { }";
+
+            Action<ModuleSymbol> validator = module =>
+            {
+                var typeParameter = module.GlobalNamespace.GetTypeMember("Test").TypeParameters.Single();
+                Assert.False(typeParameter.HasValueTypeConstraint);
+                Assert.False(typeParameter.HasReferenceTypeConstraint);
+                Assert.True(typeParameter.HasConstructorConstraint);
+                Assert.Equal(SpecialType.System_Enum, typeParameter.ConstraintTypes().Single().SpecialType);
+            };
+
+            CompileAndVerify(code, sourceSymbolValidator: validator, symbolValidator: validator);
+        }
```

Comment:
Perhaps test a generic method override such as the following. Should `B.F<U>` include constraints for both `E` and `System.Enum` in source and metadata?
```
enum E { }
abstract class A<T>
{
    internal abstract void F<U>() where U : T, System.Enum;
}
class B : A<E>
{
    internal override void F<U>() { }
}
```

Reply:
// error CS0460: Constraints for override and explicit interface implementation methods are inherited from the base method, so they cannot be specified directly
I'll add tests as such

---
In reply to: [161892869](https://github.com/dotnet/roslyn/pull/24199#discussion_r161892869) [](ancestors = 161892869)

Reply:
Sorry, I meant should the source symbol for `B.F<U>` include both constraints, and should the metadata for `B.F<U>` include both.